docs/tasks/administer-cluster/kms-provider.md

```diff
@@ -0,0 +1,182 @@
+    - secrets
+    providers:
+    - kms:
+        name: <display name of the KMS plugin>
+        endpoint: <UNIX domain socket listen address of the gRPC server (KMS plugin)>
+        cachesize: <number of data encryption keys (DEKs) to be cached in the clear>
+   - identity: {}
+```
+
+2. Set the `--experimental-encryption-provider-config` flag on the kube-apiserver to point to the location of the configuration file.
+3. Restart your API server.
+
+## Verifying that the data is encrypted
+Data is encrypted when written to etcd. After restarting your kube-apiserver, any newly created or updated secret should be encrypted when stored. To verify, you can use the etcdctl command line program to retrieve the contents of your secret.
+
+1. Create a new secret called secret1 in the default namespace:
+```
+kubectl create secret generic secret1 -n default --from-literal=mykey=mydata
+```
+2. Using the etcdctl command line, read that secret out of etcd:
+```
+ETCDCTL_API=3 etcdctl get /kubernetes.io/secrets/default/secret1 [...] | hexdump -C
+```
+ where [...] must be the additional arguments for connecting to the etcd server.
+
+3. Verify the stored secret is prefixed with `k8s:enc:kms:v1:`, which indicates that the KMS provider has encrypted the resulting data.
+
+4. Verify that the secret is correctly decrypted when retrieved via the API:
+```
+kubectl describe secret secret1 -n default
+```
+should match `mykey: mydata`
+
+## Ensuring all secrets are encrypted
+Because secrets are encrypted on write, performing an update on a secret encrypts that content.
+
+The following command reads all secrets and then updates them to apply server side encryption. If an error occurs due to a conflicting write, retry the command. For larger clusters, you may wish to subdivide the secrets by namespace or script an update.
+```
+kubectl get secrets --all-namespaces -o json | kubectl replace -f -
+```
+
+## Switching from a local encryption provider to the KMS provider
+To switch from a local encryption provider to the KMS provider and re-encrypt all of the secrets:
+
+1. Add the KMS provider as the first entry in the configuration file as shown in the following example.
+
+```yaml
+kind: EncryptionConfig
+apiVersion: v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - kms:
+        name : <display name of the KMS plugin>
+        endpoint: <UNIX domain socket listen address of the gRPC server (KMS plugin)>
+        cachesize: <number of data encryption keys (DEKs) to be cached in the clear>
+    - aescbc:
+         keys:
+         - name: key1
+           secret: <BASE 64 ENCODED SECRET>
+```
+
+2. Restart all kube-apiserver processes.
+
+3. Run the following command to force all secrets to be re-encrypted using the KMS provider.
+
+```
+kubectl get secrets --all-namespaces -o json| kubectl replace -f -
+```
+
+## Decrypting your data
+To disable encryption at rest:
+
+1. Place the identity provider as the first entry in the configuration file: 
```

Comment:
- should identity be `identity` ?
- the config below is doing BOTH decryption and encryption disabling? This is confusing. Sounds like this section is about *disabling* encryption and *decryption* is a necessary step for turning encryption off?